mlir_graphblas/algorithms.py

```diff
@@ -2003,3 +2003,375 @@ def __call__(
+        tmp_N_values = irb.sparse_tensor.values(tmp_N)
+        z_vectors = irb.llvm.alloca(num_taus_i64, "!llvm.ptr<i8>")
+        with irb.for_loop(0, N) as z_for_vars:
+            row_index = z_for_vars.iter_var_index
+            row_index_i64 = irb.arith.index_cast(row_index, "i64")
+            tmp_N_value = irb.memref.load(tmp_N_values, row_index)
+            row = irb.graphblas.apply(range_N, "times", right=tmp_N_value)
+            row_ptr8 = irb.util.tensor_to_ptr8(row)
+            destination_ptr = irb.llvm.getelementptr(z_vectors, row_index_i64)
+            irb.llvm.store(row_ptr8, destination_ptr)
+        z = irb.util.new_sparse_tensor("tensor<?x?xf64, #CSR64>", N, N)
+        z_ptr8 = irb.util.tensor_to_ptr8(z)
+        N_squared = irb.arith.muli(N, N)
+        N_squared_i64 = irb.arith.index_cast(N_squared, "i64")
+        N_plus_one = irb.arith.addi(N, c1)
+        irb.util.resize_sparse_pointers(z_ptr8, c1, N_plus_one)
+        irb.util.resize_sparse_index(z_ptr8, c1, N_squared)
+        irb.util.resize_sparse_values(z_ptr8, N_squared)
+        z_pointers = irb.sparse_tensor.pointers(z, c1)
+        z_indices = irb.sparse_tensor.indices(z, c1)
+        z_values = irb.sparse_tensor.values(z)
+        with irb.for_loop(0, N) as z_concat_loop_row_for_vars:
+            row_index = z_concat_loop_row_for_vars.iter_var_index
+            row_index_i64 = irb.arith.index_cast(row_index, "i64")
+            n_times_row_index = irb.arith.muli(N, row_index)
+            n_times_row_index_i64 = irb.arith.index_cast(n_times_row_index, "i64")
+            irb.memref.store(n_times_row_index_i64, z_pointers, row_index)
+            row_vector_ptr = irb.llvm.getelementptr(z_vectors, row_index_i64)
+            row_vector_ptr8 = irb.llvm.load(row_vector_ptr, "!llvm.ptr<i8>")
+            row_vector = irb.util.ptr8_to_tensor(
+                row_vector_ptr8, "tensor<?xf64, #CV64>"
+            )
+            row_vector_values = irb.sparse_tensor.values(row_vector)
+            with irb.for_loop(0, N) as z_concat_loop_col_for_vars:
+                col_index = z_concat_loop_col_for_vars.iter_var_index
+                col_index_i64 = irb.arith.index_cast(col_index, "i64")
+                values_position = irb.arith.addi(n_times_row_index, col_index)
+                irb.memref.store(col_index_i64, z_indices, values_position)
+                value = irb.memref.load(row_vector_values, col_index)
+                irb.memref.store(value, z_values, values_position)
+        irb.memref.store(N_squared_i64, z_pointers, N)
+        # TODO ideally done with loop fusion with the above loops
+        z = irb.graphblas.apply(z, "cos")
+
+        c_vectors = irb.llvm.alloca(num_taus_i64, "!llvm.ptr<i8>")
+        two_over_N = irb.arith.divf(c2_f64, N_f64)
+        negative_a_over_lmax = irb.arith.divf(a, lmax)
+        negative_a_over_lmax = irb.arith.mulf(negative_a_over_lmax, c_negative_1_f64)
+        with irb.for_loop(0, num_taus) as tau_for_vars:
```

Comment:
I should add a `TODO` comment stating that this loop could be parallel. It probably won't help much if `num_taus` is small (which it is for our tests, but I'd have to check with @eriknw to see what's used in practice). If `num_taus` is small in practice, I should leave a comment stating that this is the reason it's not parallel.

Reply:
https://github.com/metagraph-dev/mlir-graphblas/commit/ce26f591efd6031e53380d12bb5641ce99a4d9bd implements this suggestion.